freqtrade/freqai/freqai_util.py

```diff
@@ -0,0 +1,115 @@
+from typing import Any, Dict, Tuple
+
+from freqtrade.configuration import TimeRange
+from freqtrade.constants import Config
+from freqtrade.exceptions import OperationalException
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_full_model_path(config: Config) -> Path:
+    """
+    Returns default FreqAI model path
+    :param config: Configuration dictionary
+    """
+    freqai_config: Dict[str, Any] = config["freqai"]
+    return Path(
+        config["user_data_dir"] / "models" / str(freqai_config.get("identifier"))
+    )
+
+
+def get_timerange_from_ready_models(models_path: Path) -> Tuple[TimeRange, str, Dict[str, Any]]:
+    """
+    Returns timerange information based on a FreqAI model directory
+    :param models_path: FreqAI model path
+
+    :returns: a Tuple with (backtesting_timerange: Timerange calculated from directory,
+    backtesting_string_timerange: str timerange calculated from
+    directory (format example '20020822-20220830'), \
+    pairs_end_dates: Dict with pair and model end training dates info)
+    """
+    all_models_end_dates = []
+    pairs_end_dates: Dict[str, Any] = get_pairs_timestamps_training_from_ready_models(models_path)
+    for key in pairs_end_dates:
+        for model_end_date in pairs_end_dates[key]:
+            if model_end_date not in all_models_end_dates:
+                all_models_end_dates.append(model_end_date)
+
+    if len(all_models_end_dates) == 0:
+        raise OperationalException(
+            'At least 1 saved model is required to '
+            'run backtest with the freqai-backtest-live-models option'
+        )
+
+    if len(all_models_end_dates) == 1:
+        logger.warning(
+            "Only 1 model was found. Backtesting will run with the "
+            "timerange from the end of the training date to the current date"
+        )
+
+    finish_timestamp = int(datetime.now(tz=timezone.utc).timestamp())
+    if len(all_models_end_dates) > 1:
+        # After last model end date, use the same period from previous model
+        # to finish the backtest
+        all_models_end_dates.sort(reverse=True)
+        finish_timestamp = all_models_end_dates[0] + \
+            (all_models_end_dates[0] - all_models_end_dates[1])
+
+    all_models_end_dates.append(finish_timestamp)
+    all_models_end_dates.sort()
+    start = datetime.fromtimestamp(min(all_models_end_dates), tz=timezone.utc)
+    stop = datetime.fromtimestamp(max(all_models_end_dates), tz=timezone.utc)
+    end_date_string_timerange = stop
+    if (
+        finish_timestamp < int(datetime.now(tz=timezone.utc).timestamp()) and
+        datetime.now(tz=timezone.utc).strftime('%Y%m%d') != stop.strftime('%Y%m%d')
+    ):
+        # add 1 day to string timerange to ensure BT module will load all dataframe data
+        end_date_string_timerange = stop + timedelta(days=1)
+
+    backtesting_string_timerange = (
+        f"{start.strftime('%Y%m%d')}-{end_date_string_timerange.strftime('%Y%m%d')}"
+    )
```

Comment:
I'm not sure on the usefulness of this.

While it's currently used for backtesting - that part could easily be done by using the "TimeRange" object from below in combination with an else.(config.timerange should not be used elsewhere in backtesting).

It's also assigned to `backtest_live_models_data` - but this part / entry is never used - so i think we can save the 2nd value (as well as the conversion to that).

Reply:
I refactored the code, but `config.timerange` is used in freqai. I think that for now, we need this config. Example  below:

https://github.com/freqtrade/freqtrade/blob/develop/freqtrade/freqai/data_kitchen.py#L88

Reply:
mmh i see - nice catch :+1: